ujt/state.py

```diff
@@ -66,43 +67,194 @@ def read_local_data() -> Tuple[Dict[str, Node], Dict[str, Client]]:
+    
+    Args:
+        virutal_node_map: The new virtual node map to be saved in the cache.
+    """
+    cache.set("virtual_node_map", virtual_node_map)
+
+
+def get_parent_virtual_node_map() -> Dict[str, str]:
+    """ Gets a dictionary mapping node names to the name of their direct virtual node parent.
+
+    The keys of the dictionary can be names of virtual and non-virtual nodes.
+    The values are always virtual nodes. 
+    This dictionary can be used to re-construct the chain of the virtual nodes that contain a given node. 
+
+    Returns:
+        A dictionary mapping node names to the name of their direct virtual node parent.
+    """
+    return cache.get("parent_virtual_node_map")
+
+
+def set_parent_virtual_node_map(parent_virtual_node_map: Dict[str, str]):
+    """ Sets a dictionary mapping node names to the name of their direct virtual node parent.
+    
+    Args:
+        parent_virutal_node_map: The new parent virtual node map to be saved in the cache.
+    """
+    cache.set("parent_virtual_node_map", parent_virtual_node_map)
+
+
+def add_virtual_node(
+    virtual_node_name: str,
+    selected_node_data: List[Dict[str,
+                                  Any]],
+):
+    """ Adds a virtual node.
+
+    Updates the virtual node map with the newly created virtual node object.
+    Updates the entries in the parent virtual node map corresponding to the virtual node's children,
+    to point to the new virtual node.
+
+    The interface could be refactored to take node_names instead of selected_node_data to be cleaner.
+    However, this function is currently only called in the callback to update the elements of the cytoscape graph.
+    It would be an extraneous transformation that doesn't offer any additional convenience or benefit, currently.
+
+    Args:
+        virutal_node_name: The name of the virtual node to create.
+        selected_node_data: A list of node data dictionaries to include in the virtual node.
+
+    """
+    virtual_node_map = get_virtual_node_map()
+    parent_virtual_node_map = get_parent_virtual_node_map()
+    node_name_message_map = get_node_name_message_map()
+
+    virtual_node_child_names: Set[str] = set()
+    # use this queue to do BFS to flatten non-virtual nodes
+    node_frontier: Deque[Node] = deque()
+    for node_data in selected_node_data:
+        if node_data["ujt_id"] in virtual_node_map:  # nested virtual node
+            virtual_node_child_names.add(node_data["ujt_id"])
+            parent_virtual_node_map[node_data["ujt_id"]] = virtual_node_name
+        else:
+            node_frontier.append(node_name_message_map[node_data["ujt_id"]])
+
+    while node_frontier:
+        node = node_frontier.popleft()
+        virtual_node_child_names.add(node.name)
+        parent_virtual_node_map[node.name] = virtual_node_name
+        for child_name in node.child_names:
+            node_frontier.append(node_name_message_map[child_name])
```

Comment:
Here's the code to create a new virtual node. We keep track of a virtual node's children. This parent/child relationship is subtly different from those between actual nodes, as this relationship isn't reflected in the Node proto itself. It's only held on the virtual node's end. Moreover, if a virtual node X contains Service1, Service1.Endpoint1 and Service1.Endpoint2, X's children are all the aforementioned nodes. (They are collapsed).